lib/parachain/networkprotocol.go

```diff
@@ -0,0 +1,191 @@
+	Validator ValidatorIndex `scale:"2"`
+	// Cert the cert itself.
+	Cert AssignmentCert `scale:"3"`
+}
+
+// CandidateIndex the index of the candidate in the list of candidates fully included as-of the block.
+type CandidateIndex uint32
+
+// Assignment holds indirect assignment cert and candidate index
+type Assignment struct {
+	IndirectAssignmentCert IndirectAssignmentCert `scale:"1"`
+	CandidateIndex         CandidateIndex         `scale:"2"`
+}
+
+// Assignments for candidates in recent, unfinalized blocks.
+type Assignments struct {
+	Assignments []Assignment
+}
+
+// Index returns varying data type index
+func (a Assignments) Index() uint {
+	return 0
+}
+
+// ValidatorSignature with which parachain validators sign blocks.
+type ValidatorSignature [sr25519.SignatureLength]byte
+
+// IndirectSignedApprovalVote A signed approval vote which references the candidate indirectly via the block.
+//
+// In practice, we have a look-up from block hash and candidate index to candidate hash,
+// so this can be transformed into a `SignedApprovalVote`.
+type IndirectSignedApprovalVote struct {
+	// BlockHash a block hash where the candidate appears.
+	BlockHash common.Hash `scale:"1"`
+	// CandidateIndex the index of the candidate in the list of candidates fully included as-of the block.
+	CandidateIndex CandidateIndex `scale:"2"`
+	// ValidatorIndex the validator index.
+	ValidatorIndex ValidatorIndex `scale:"3"`
+	// Signature the signature of the validator.
+	Signature ValidatorSignature `scale:"4"`
+}
+
+// Approvals for candidates in some recent, unfinalized block.
+type Approvals struct {
+	Approvals []IndirectSignedApprovalVote
+}
+
+// Index returns varying data type index
+func (ms Approvals) Index() uint {
+	return 1
+}
+
+// ApprovalDistributionMessage network messages used by approval distribution subsystem.
+type ApprovalDistributionMessage scale.VaryingDataType
+
+// Set will set a VoryingDataTypeValue using the underlying VaryingDataType
+func (adm *ApprovalDistributionMessage) Set(val scale.VaryingDataTypeValue) (err error) {
+	vdt := scale.VaryingDataType(*adm)
+	err = vdt.Set(val)
+	if err != nil {
+		return fmt.Errorf("setting value to varying data type: %w", err)
+	}
+	*adm = ApprovalDistributionMessage(vdt)
+	return nil
+}
+
+// Value returns the value from the underlying VaryingDataType
+func (adm *ApprovalDistributionMessage) Value() (scale.VaryingDataTypeValue, error) {
+	vdt := scale.VaryingDataType(*adm)
+	return vdt.Value()
+}
+
+// New returns new ApprovalDistributionMessage VDT
+func (adm ApprovalDistributionMessage) New() ApprovalDistributionMessage {
+	return NewApprovalDistributionMessageVDT()
+}
+
+// NewApprovalDistributionMessageVDT ruturns a new ApprovalDistributionMessage VaryingDataType
+func NewApprovalDistributionMessageVDT() ApprovalDistributionMessage {
+	vdt, err := scale.NewVaryingDataType(Assignments{}, Approvals{})
+	if err != nil {
+		panic(err)
+	}
+	return ApprovalDistributionMessage(vdt)
+}
```

Comment:
I believe we need both,  `func (adm ApprovalDistributionMessage) New() ApprovalDistributionMessage` is necessary because scale will uses this function to decode nested VDTs.  When I try to instantiate Approval Distribution Message with  `message := ApprovalDistributionMessage.New()` I get error: 
```
lib/parachain/approval_distribution_message_test.go:18:33: not enough arguments in call to ApprovalDistributionMessage.New
        have ()
        want (ApprovalDistributionMessage) (typecheck)

```
So I need to use: `message := NewApprovalDistributionMessageVDT()`

Let me know if there is a different way to approach this.

Reply:
I understand it now. I think it is okay the way it is right now.